### ISSUE TITLE
(Refactoring) Process files using InputStreams, not arrays of lines.

### ISSUE THREAD

**Issue by OndraZizka** [jbake-org/jbake#493]
Currently, JBake reads files to a List of lines and then processing happens line-by-line.
That has few gotchas.
* It is not too performant
* The whole file is read to memory
* We can't use the multitude of existing Input/OutputStream utils (like [BOMInputStream](https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/input/BOMInputStream.html)
* If we hit a large binary file by mistake, it may end up as a single line and then fed to things like .contains(<bom>)

So I suggest to replace this line-by-line approach with Input/OutputStreams.
* Streams are easier to parallelize
* Streams are easier to reuse (e.g.: read until header.separator into header stream, read from that to the end into a body stream)

Perhaps something for 3.0?

**Comment by ancho** [comment]
Yes. That's a good idea. There is much more potential in optimizing things.
We are currently throwing lots of hashmaps around, creating deep copies with full content of documents in-memory. We could parralelize rendering things and stuff like that.

Working with streams is a first good step. Don't know if waiting for 3.0 is a good idea.
On the other hand, we could start a 3.0 branch throw java 7 to the bin and start using better language features.

**Comment by OndraZizka** [comment]
I am all for switching to Java 8. Even conservative businesses go Java 8 now because streams (the other streams :) are so tempting, and therefore many libs and frameworks switch too.

+1 to turning hashmaps shuffling to normal OOP. The current state makes coding a bit harder (for me at least).

Parallelization is a good thing but I would postpone that, since the performance is secondary concern, I was quite surprised that it takes around 10 seconds to convert all my files including calling the HTTP service :)

**Comment by ancho** [comment]
> Parallelization is a good thing but I would postpone that, since the performance is secondary concern, I was quite surprised that it takes around 10 seconds to convert all my files including calling the HTTP service :)

Jep. First make it stable then make it fast!

> I'm not sure if that would make a good issue, "Change hashmaps to classes", that would be quite a lot of work. So perhaps - case by case, when someone does a PR?

That's a good question. I was thinking about controlled abstraction for the different models we have.
document models, render/template models or what ever we got there.

So refactor template models could be a good start.

> +1 to turning hashmaps shuffling to normal OOP. The current state makes coding a bit harder (for me at least).

It's still difficult from time to time for me as well. Hard to remember when, where and what changes the "state" of the map. And is it a configuration map, a document map or a template map?
(One reason I started my big refactoring session of the configuration stuff back then)

But one lesson I learned. To many changes at once makes a review extremely difficult. It's one year ago and still not merged. Which makes it more and more tedious to rebase with master.
